<a href="https://colab.research.google.com/github/ujeongmin/ESAA/blob/main/ESAA_HW_9_11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## [04. 데이터 핸들링 - 판다스]

*   데이터 처리를 위해 존재하는 가장 인기 있는 라이브러리
*   대부분의 데이터 세트는 행x열로 구성된 2차원 데이터이며, 이를 가장 효율적으로 가공/처리할 수 있는 다양하고 훌륭한 기능 제공
*   넘파이에 비해 훨씬 유연하고 편리하게 데이터 핸들링 가능
*   핵심 객체인 DataFrame



---

### 판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API
*   판다스는 pd로 표현

In [1]:
import pandas as pd

* titanic 데이터 불러오기

In [9]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


* head(N)은 맨 앞에있는 N개의 로우를 반환

In [11]:
titanic_df = pd.read_csv('/content/gdrive/MyDrive/ESAA/data/titanic_train.csv')
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [12]:
print('titanic 변수 type:', type(titanic_df))

titanic 변수 type: <class 'pandas.core.frame.DataFrame'>


In [13]:
print('DataFrame 크기:', titanic_df.shape)

DataFrame 크기: (891, 12)


* titanic_df는 891개의 로우와 12개의 칼럼

In [14]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


info()를 이용하여 총 데이터 건수와 데이터 타입, Null 건수를 알 수 있다
* RangeIndex : DataFrame index의 범위를 나타냄, 전체 로우와 칼럼 수 알 수 있음
* 칼럼별 데이터 타입 : ex) Pclass 칼럼의 데이터 타입은 int64형, Name 칼럼은 object(문자열 타입)
* nnn non-null : 몇개의 데이터가 non-null인지 나타냄, 전체 891개에서 nnn개가 non-null이고 나머지는 null
* 아래에 dtypes : 전체 12개의 칼럼들의 타입이 요약되어 있음

In [15]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

describe()를 이용하여 숫자형 칼럼들에 대한 개략적인 데이터 분포도를 확인할 수 있다  
* PassengerID 칼럼은 승객 ID 식별 칼럼이므로 의미 있는 속성 아님
* Survived는 0,1로 이루어진 숫자형 카테고리 칼럼
* Pclass도 1,2,3으로 이루어진 숫자형 카테고리 칼럼

In [16]:
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)

3    491
1    216
2    184
Name: Pclass, dtype: int64


value_counts()는 많은 건수로 정렬되어 값을 반환
* Pclass값 3이 491개, 1이 216개, 2가 184개라는 뜻

In [18]:
titanic_pclass = titanic_df['Pclass']
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


Dataframe[ 칼럼명 ]
* Series는 Index와 단 하나의 칼럼으로 구성된 데이터 세트

In [19]:
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

* 왼쪽은 Series의 Index, 오른쪽은 데이터 값

In [20]:
value_counts = titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
3    491
1    216
2    184
Name: Pclass, dtype: int64


---

### DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환
* 기본적으로 대부분 넘파이 ndarray를 입력 인자로 사용함



####넘파이 ndarray, 리스트, 딕셔너리를 Dataframe으로 변환하기
* 1차원 형태의 리스트와 ndarray를 DataFrame으로 변환하기

In [21]:
import numpy as np

col_name1 = ['col1']
list1 = [1,2,3]
array1 = np.array(list1)
print('array1 shape:', array1.shape)
# 리스트를 이용해 DataFrame 생성
df_list1 = pd.DataFrame(list1, columns = col_name1)
print('1차원 리스트로 만든 DataFrame:\n', df_list1)
# 넘파이 ndarray를 이용해 dataframe 생성
df_array1 = pd.DataFrame(array1, columns = col_name1)
print('1차원 ndarray로 만든 DataFrame:\n', df_array1)

array1 shape: (3,)
1차원 리스트로 만든 DataFrame:
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame:
    col1
0     1
1     2
2     3


* 2차원 형태의 리스트와 ndarray를 DataFrame으로 변환하기

In [22]:
# 3개의 칼럼명이 필요함 (3열 리스트이므로)
col_name2 = ['col1','col2','col3']

# 2행x3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환
list2 = [[1,2,3],
         [11,12,13]]
array2 = np.array(list2)
print('array2 shape:', array2.shape)
# 리스트를 이용해 DataFrame 생성
df_list2 = pd.DataFrame(list2, columns = col_name2)
print('2차원 리스트로 만든 DataFrame:\n', df_list2)
# 넘파이 ndarray를 이용해 dataframe 생성
df_array2 = pd.DataFrame(array2, columns = col_name2)
print('2차원 ndarray로 만든 DataFrame:\n', df_array2)

array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13


* 딕셔너리를 DataFrame으로 변환하기

In [23]:
# Key는 문자열 칼럼명으로 매핑, Value는 리스트 형(또는 ndarray) 칼럼 데이터로 매핑
dict = {'col1':[1,11], 'col2':[2,22], 'col3':[3,33]}
df_dict = pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame:\n', df_dict)

딕셔너리로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    22    33


* DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기

In [24]:
# DataFrame을 ndarray로 변환
array3 = df_dict.values
print('df_dict.values 타입:', type(array3), 'df_dict.values shape:', array3.shape)
print(array3)

df_dict.values 타입: <class 'numpy.ndarray'> df_dict.values shape: (2, 3)
[[ 1  2  3]
 [11 22 33]]


In [25]:
# DataFrame을 리스트로 변환
list3 = df_dict.values.tolist()
print('df_dict.values.tolist() 타입:', type(list3))
print(list3)

# DataFrame을 딕셔너리로 변환
dict3 = df_dict.to_dict('list')
print('\n df_dict.to_dict() 타입:', type(dict3))
print(dict3)

df_dict.values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to_dict() 타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


---

### DataFrame의 칼럼 데이터 세트 생성과 수정
* [ ]연산자를 이용해 쉽게 가능, [ ] 내에 새로운 칼럼명을 입력하고 값을 할당해주기만 하면 됨

In [26]:
titanic_df['Age_0'] = 0
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

* 새로운 칼럼명 Age_0 으로 모든 데이터값이 0으로 할당된 Series가 기존 DataFrame에 추가됨

In [27]:
titanic_df['Age_by_10'] = titanic_df['Age']*10
titanic_df['Family_No'] = titanic_df['SibSp'] + titanic_df['Parch'] + 1
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          2  
1          2  
2          1

* 새로운 칼럼 Series인 Age_by_10과 Family_no가 새롭게 DataFrame에 추가됨

기존 칼럼 값들도 쉽게 일괄적으로 업데이터 가능
* 업데이트를 원하는 칼럼 Series를 [ ] 내에 칼럼명으로 입력한 뒤 값 할당해주면 됨

In [28]:
titanic_df['Age_by_10'] = titanic_df['Age_by_10'] + 100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

---

### DataFrame 데이터 삭제
* drop()을 이용
> DataFrame.drop(labels = None, axis = 0, index = None, columns = None, level = None, inplace = False, errors = 'raise)

* 가장 중요한 파라미터는 labels, axis, inplace
> axis 값에 따라서 특정 로우(axis=0) 또는 칼럼(axis=1) 을 드롭  
> labels에 원하는 칼럼 명 입력하고 axis=1을 입력하면 지정된 칼럼을 드롭  
> axis=0일 경우 labels에 오는 값을 인덱스로 간주

In [29]:
titanic_drop_df = titanic_df.drop('Age_0', axis = 1)
titanic_drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          2  
1      0          PC 17599  71.2833   C85        C      480.0          2  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          1

* Age_0 칼럼이 삭제됨

In [30]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

* 하지만 원본 titanic DataFrame을 보면 Age_0 칼럼이 여전히 존재함
> 이는 Inplace = False 때문임  
> 자기 자신의 DataFrame의 데이터는 삭제하지 않고 삭제된 결과만을 반환하는 것  
> True로 설정하면 자기 자신도 삭제됨
* 여러 칼럼을 삭제하고 싶으면 리스트 형태로 labels에 입력

In [31]:
drop_result = titanic_df.drop(['Age_0','Age_by_10','Family_No'], axis=1, inplace=True)
print('inplace=True 로 drop 후 반환된 값:', drop_result)
titanic_df.head(3)

inplace=True 로 drop 후 반환된 값: None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

* inplace = True로 설정 시 반환값이 None이 되는 것 주의  
* 그렇기에 titanic_df = titanic_df.drop(inplace=True) 이렇게 다시 자신의 원래 DataFrame 객체로 할당하면 안됨
* drop_titanic 등 새로운 객체에 할당해줘야 함

로우 삭제

In [33]:
pd.set_option('display.width',1000)
pd.set_option('display.max_colwidth',15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))
titanic_df.drop([0,1,2], axis=0, inplace=True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


* Index 0,1,2에 위치한 로우가 삭제됨

* drop() 정리
> axis : 로우 삭제는 0, 칼럼 삭제는 1  
> 원본 유지하고 새로운 객체로 받고 싶다면 inplace = False  
> 원본에 드롭된 결과를 적용하고 싶다면 inplace = True  


---

### Index 객체
* 판다스의 Index 객체는 PK와 유사하게 DataFrame, Series의 레코드를 고유하게 식별하는 객체

In [35]:
# 원본 파일 다시 로딩
titanic_df = pd.read_csv('/content/gdrive/MyDrive/ESAA/data/titanic_train.csv')
# Index 객체 추출
indexes = titanic_df.index
print(indexes)
# Index 객체를 실제 값 array로 변환
print('Index 객체 array값:\n', indexes.values)

RangeIndex(start=0, stop=891, step=1)
Index 객체 array값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232

* Index 객체는 식별성 데이터를 1차원 array로 가지고 있음
* ndarray와 유사하게 단일 값 반환 및 슬라이싱도 가능

In [37]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


* 한 번 만들어진 DataFrame 및 Series의 Index 객체는 함부로 변경할 수 없음

In [38]:
indexes[0] = 5

TypeError: ignored

* Series 객체는 Index 객체를 포함하지만 연산에서는 제외됨, 오직 식별용

In [40]:
series_fair = titanic_df['Fare']
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series:', sum(series_fair))
print('Fair Series + 3:\n', (series_fair+3).head(3))

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series: 28693.949299999967
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


* DataFrame 및 Series에 reset_index()를 수행하면 새롭게 인덱스를 연속 숫자 형으로 할당
* 기존 인덱스는 'index'라는 새로운 칼럼 명으로 추가

In [41]:
titanic_reset_df = titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

* reset_index(drop = True) 로 설정하면 기존 인덱스가 새로운 칼럼으로 추가되지 않고 삭제됨

In [42]:
print('### before reset_index ###')
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입:', type(value_counts))
new_value_counts = value_counts.reset_index(inplace=False)
print('### after reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:', type(new_value_counts))

### before reset_index ###
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
### after reset_index ###
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


---

### 데이터 셀렉션 및 필터링
#### DataFrame의 [ ] 연산자
* 넘파이와 DataFrame 간 데이터 셀렉션에서 가장 유의해야 할 부분임
> 넘파이에서 [ ] 연산자는 행/열의 위치, 슬라이싱 범위 등을 지정해 데이터 가져옴  
> DataFrame 바로 뒤의 [ ] 안에는 '칼럼명 문자' 또는 인덱스로 변환 가능한 표현식이 들어옴

In [43]:
print('단일 칼럼 데이터 추출:\n', titanic_df['Pclass'].head(3))
print('\n여러 칼럼의 데이터 추출:\n', titanic_df[['Survived','Pclass']].head(3))
print('[ ] 안에 숫자 index는 KeyError 오류 발생:\n', titanic_df[0])

단일 칼럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: ignored

* 위에선 [ ] 내에 숫자 값 입력하면 오류가 발생했으나 슬라이싱, 불린 인덱싱과 같은 인덱스 형태로 변환 가능한 표현식은 입력 가능함

In [44]:
titanic_df[0:2]         # 슬라이싱, 가능하긴 하나 혼동을 주기에 사용하지 않는 것이 좋음

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0  A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0   PC 17599  71.2833   C85        C

In [45]:
titanic_df[titanic_df['Pclass']==3].head(3)       # 불린 인덱싱

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket   Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171  7.250   NaN        S
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...  7.925   NaN        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450  8.050   NaN        S

#### DataFrame ix[ ] 연산자
* 넘파이 ndarray의 [ ] 연산자와 유사한 기능을 DataFrame에 제공하기 위해서 ix[ ]를 제공함
* 예를 들어 ix[0,'Pclass']는 인덱스 0인 로우의 Pclass 칼럼, ix[0,2]는 인덱스 0인 로우의 세 번째 칼럼
* ix[ ]는 잘 사용되고 있으나 향후 사라질 수 있으니 대체 가능한 loc[ ] 또는 iloc[ ]을 권장함

In [70]:
# print('칼럼 위치 기반 인덱싱 데이터 추출:', titanic_df.ix[0,2])
# print('칼럼 명 기반 인덱싱 데이터 추출:', titanic_df.ix[0,'Pclass'])

In [64]:
data = {'Name':['Chulmin','Eunkyung', 'Jinwoong','Soobeom'],
        'Year':[2011,2016,2015,2015],
        'Gender':['Male','Female','Male','Male']}
data_df = pd.DataFrame(data, index = ['one','two','three','four'])
data_df

Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

#### 명칭 기반 인덱싱과 위치 기반 인덱싱의 구분
* 명칭 기반 인덱싱은 칼럼의 명칭으로 위치 지정/위치 기반은 0부터 시작하는 위치를 기반으로 지정

In [51]:
# data_df를 reset_index()로 새로운 숫자형 인덱스를 생성
data_df_reset = data_df.reset_index()
data_df_reset = data_df_reset.rename(columns={'index':'old_index'})

# 인덱스값에 1을 더해서 1부터 시작하는 새로운 인덱스값 생성
data_df_reset.index = data_df_reset.index+1
data_df_reset

old_index      Name  Year  Gender
1       one   Chulmin  2011    Male
2       two  Eunkyung  2016  Female
3     three  Jinwoong  2015    Male
4      four   Soobeom  2015    Male

In [69]:
# data_df_reset.ix[1,1]

#### DataFrame iloc[ ] 연산자
* iloc[ ]는 위치 기반 인덱싱만 허용하므로 행과 열 값으로 integer 형의 슬라이싱, 팬시 리스트 값을 입력해줘야 함

In [50]:
data_df.iloc[0,0]    # 첫번째 행, 첫번째 열

'Chulmin'

* 위치 인덱싱이 아닌 명칭을 입력하면 오류 발생

In [53]:
# 다음 코드는 오류를 발생합니다
data_df.iloc[0,'Name']

ValueError: ignored

In [55]:
# 다음 코드는 오류를 발생합니다
data_df.iloc['one',0]

ValueError: ignored

In [57]:
# 위치 기반 인덱싱은 잘 반환
data_df_reset.iloc[0,1]

'Chulmin'

#### DataFrame loc[ ] 연산자
* 명칭 기반으로 데이터 추출

In [58]:
data_df.loc['one','Name']

'Chulmin'

* index가 숫자형일 수 있기 때문에 명칭 기반이라고 무조건 문자열을 입력한다는 것은 아님

In [59]:
data_df_reset.loc[1,'Name']

'Chulmin'

In [60]:
# 인덱스 0인 행이 없으므로 오류 발생
data_df_reset.loc[0,'Name']

KeyError: ignored

* loc[ ]에 슬라이싱 기호 ':'를 적용할때 유의할 점
> 종료 값-1이 아니라 종료 값까지 포함함  
> 즉 0:3이면 0,1,2,3을 의미함

In [66]:
# print('명칭 기반 ix slicing\n', data_df.ix['one':'two':'Name'], '\n')
print('위치 기반 iloc slicing\n', data_df.iloc[0:1,0], '\n')
# print('명칭 기반 loc slicing\n', data_df.loc['one':'two':'Name'])

위치 기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 



In [68]:
print(data_df_reset.loc[1:2,'Name'])

1     Chulmin
2    Eunkyung
Name: Name, dtype: object


In [74]:
# print(data_df.ix[1:2,'Name'])

#### 불린 인덱싱
* ix, iloc, loc 보다는 불린 인덱싱이 매우 편리하고 많이 쓰임

In [75]:
titanic_df = pd.read_csv('/content/gdrive/MyDrive/ESAA/data/titanic_train.csv')
titanic_boolean = titanic_df[titanic_df['Age']>60]
print(type(titanic_boolean))
titanic_boolean

<class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked
33            34         0       2  Wheadon, Mr...    male  66.0      0      0   C.A. 24579   10.5000          NaN        S
54            55         0       1  Ostby, Mr. ...    male  65.0      0      1       113509   61.9792          B30        C
96            97         0       1  Goldschmidt...    male  71.0      0      0     PC 17754   34.6542           A5        C
116          117         0       3  Connors, Mr...    male  70.5      0      0       370369    7.7500          NaN        Q
170          171         0       1  Van der hoe...    male  61.0      0      0       111240   33.5000          B19        S
252          253         0       1  Stead, Mr. ...    male  62.0      0      0       113514   26.5500          C87        S
275          276         1       1  Andrews, Mi...  female  63.0      1      0        13502   77.9583           D7        S
280          281         0       3  Duane, Mr. ...    male  65.0      0      0       336439    7.7500          NaN        Q
326          327         0       3  Nysveen, Mr...    male  61.0      0      0       345364    6.2375          NaN        S
438          439         0       1  Fortune, Mr...    male  64.0      1      4        19950  263.0000  C23 C25 C27        S
456          457         0       1  Millet, Mr....    male  65.0      0      0        13509   26.5500          E38        S
483          484         1       3  Turkula, Mr...  female  63.0      0      0         4134    9.5875          NaN        S
493          494         0       1  Artagaveyti...    male  71.0      0      0     PC 17609   49.5042          NaN        C
545          546         0       1  Nicholson, ...    male  64.0      0      0          693   26.0000          NaN        S
555          556         0       1  Wright, Mr....    male  62.0      0      0       113807   26.5500          NaN        S
570          571         1       2  Harris, Mr....    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S
625          626         0       1  Sutton, Mr....    male  61.0      0      0        36963   32.3208          D50        S
630          631         1       1  Barkworth, ...    male  80.0      0      0        27042   30.0000          A23        S
672          673         0       2  Mitchell, M...    male  70.0      0      0   C.A. 24580   10.5000          NaN        S
745          746         0       1  Crosby, Cap...    male  70.0      1      1    WE/P 5735   71.0000          B22        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0       113572   80.0000          B28      NaN
851          852         0       3  Svensson, M...    male  74.0      0      0       347060    7.7750          NaN        S

* Age 칼럼 값이 60보다 큰 데이터를 모두 반환

In [76]:
titanic_df[titanic_df['Age']>60][['Name','Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

* Age가 60세 이상인 승객의 나이와 이름만 추출

In [77]:
titanic_df.loc[titanic_df['Age']>60, ['Name','Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

* loc [ ]을 이용해도 동일하게 적용 가능  

조건 연산자
1.   and는 &
2.   or 은 |
3.   Not은 ~



In [79]:
titanic_df[(titanic_df['Age']>60) & (titanic_df['Pclass']==1) & (titanic_df['Sex']=='female')]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

In [81]:
cond1 = titanic_df['Age']>60
cond2 = titanic_df['Pclass']==1
cond3 = titanic_df['Sex']=='female'
titanic_df[cond1 & cond2 & cond3]           # 이런 식도 가능함

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

---

### 정렬, Aggregation 함수, GroupBy 적용
#### DataFrame, Series의 정렬 - sort_values()
* sort_values()는 SQL의 order by 키워드와 매우 유사함
* 주요 파라미터는 by, ascending, inplace
> by : 해당 칼럼으로 정렬 수행  
> ascending = True : 오름차순(default)  
> inplace = False : 기존 DataFrame 유지, 정렬된 결과만 반환

In [85]:
titanic_sorted = titanic_df.sort_values(by=['Name'])         # 이름 오름차 순으로 정렬됨
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked
845          846         0       3  Abbing, Mr....    male  42.0      0      0  C.A. 5547   7.55   NaN        S
746          747         0       3  Abbott, Mr....    male  16.0      1      1  C.A. 2673  20.25   NaN        S
279          280         1       3  Abbott, Mrs...  female  35.0      1      1  C.A. 2673  20.25   NaN        S

In [84]:
titanic_sorted = titanic_df.sort_values(by=['Pclass','Name'],ascending=False)                  # 내림차 순으로 정렬됨
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch    Ticket  Fare Cabin Embarked
868          869         0       3  van Melkebe...  male   NaN      0      0    345777   9.5   NaN        S
153          154         0       3  van Billiar...  male  40.5      0      2  A/5. 851  14.5   NaN        S
282          283         0       3  de Pelsmaek...  male  16.0      0      0    345778   9.5   NaN        S

#### Aggregation 함수 적용
* min, max, sum, count와 같은 aggregation 함수도 SQL과 유사
* 모든 칼럼에 해당 aggregation을 적용함

In [86]:
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

* 특정 칼럼에 적용하기 위해서는 DataFrame에 대상 칼럼들만 추출해 적용하면 됨

In [87]:
titanic_df[['Age','Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

#### groupby() 적용

In [88]:
titanic_groupby = titanic_df.groupby(by='Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


* Pclass 칼럼 기준으로 groupby된 객체 반환

In [89]:
titanic_groupby = titanic_df.groupby('Pclass').count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked
Pclass                                                                                    
1               216       216   216  216  186    216    216     216   216    176       214
2               184       184   184  184  173    184    184     184   184     16       184
3               491       491   491  491  355    491    491     491   491     12       491

In [90]:
titanic_groupby = titanic_df.groupby('Pclass')['PassengerId','Survived'].count()
titanic_groupby

<ipython-input-90-dba2e50128e1>:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  titanic_groupby = titanic_df.groupby('Pclass')['PassengerId','Survived'].count()


PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

* Pclass로 groupby된 객체에 PassengerId와 Survived 칼럼에만 count() 수행

In [91]:
titanic_df.groupby('Pclass')['Age'].agg([max,min])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

* 서로 다른 여러 aggregation 함수를 적용할 경우에는 agg() 내에 인자로 입력해서 사용함

In [92]:
agg_format = {'Age':'max','SibSp':'sum','Fare':'mean'}
titanic_df.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

* 딕셔너리를 이용해 Pclass 별 Age의 max, SibSp의 sum, Fare의 mean 구한 것

---

### 결손 데이터 처리하기
* 결손 데이터 NULL = 넘파이에서는 NaN
* NaN 여부를 확인하는 API는 isna()
* NaN 값을 다른 값으로 대체하는 API는 fillna()

#### isna()로 결손 데이터 여부 확인
* NaN인지 아닌지를 True나 False로 알려줌

In [93]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  False  False  False  False   False  False   True     False
1        False     False   False  False  False  False  False  False   False  False  False     False
2        False     False   False  False  False  False  False  False   False  False   True     False

In [94]:
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

* sum()함수로 결손 데이터의 개수를 구할 수 있음
* 내부적으로 True는 1, False는 0으로 변환되므로 가능함

#### fillna()로 결손 데이터 대체하기


In [95]:
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500  C000        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250  C000        S

* Cabin 칼럼의 NaN 값을 C000으로 대체함
* 여기도 inplace = True 같이 실제 데이터 세트 값 바뀌는지 주의해야 함

In [96]:
titanic_df['Age'] = titanic_df['Age'].fillna(titanic_df['Age'].mean())
titanic_df['Embarked'] = titanic_df['Embarked'].fillna('S')
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

* Age 칼럼의 NaN를 평균 나이로, Embarked 칼럼의 NaN를 'S'로 대체하여 모든 결손 데이터 처리함

---

### apply lambda 식으로 데이터 가공
* 복잡한 데이터 가공이 필요할 경우 apply lambda 이용
* 함수형 프로그래밍에 사용됨

In [97]:
def get_square(a):
    return a**2

print('3의 제곱은:', get_square(3))

3의 제곱은: 9


* 원래는 get_square(a):와 같이 함수명과 입력 인자를 먼저 선언하고 결과값을 return으로 반환해야 함

In [99]:
lambda_square = lambda x : x ** 2
print('3의 제곱은:', lambda_square(3))

3의 제곱은: 9


* lambda는 이러한 것들을 한 줄로 쉽게 변환하는 식

In [100]:
a = [1,2,3]
squares = map(lambda x : x**2, a)
list(squares)

[1, 4, 9]

* 여러 개의 값을 입력 인자로 사용할 경우에는 map()함수를 결합해서 사용함

In [101]:
titanic_df['Name_len'] = titanic_df['Name'].apply(lambda x : len(x))
titanic_df[['Name', 'Name_len']].head(3)

Name  Name_len
0  Braund, Mr....        23
1  Cumings, Mr...        51
2  Heikkinen, ...        22

* Name 칼럼의 문자열 개수를 Name_len 칼럼으로 생성

In [102]:
titanic_df['Child_Adult'] = titanic_df['Age'].apply(lambda x : 'Child' if x <= 15 else 'Adult')
titanic_df[['Age', 'Child_Adult']].head(8)

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

* if else를 지원함
> 주의할 점은 if절의 경우 if 식보다 반환 값을 먼저 기술해야 함  
> ex) lambda : 'Child' if x <= 15 else 'Adult
* else if는 지원하지 않음

In [103]:
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : 'Child' if x <= 15 else ('Adult' if x <= 60 else 'Elderly'))
titanic_df['Age_cat'].value_counts()

Adult      786
Child       83
Elderly     22
Name: Age_cat, dtype: int64

* 15세 이하는 Child, 15~60세는 Adult, 61세 이상은 Elderly로 분류하는 Age_Cat 칼럼 생성

In [104]:
# 나이에 따라 세분화된 분류를 수행하는 함수 생성
def get_category(age):
  cat = ''
  if age <= 5: cat = 'Baby'
  elif age <= 12: cat = 'Child'
  elif age <= 18: cat = 'Teenager'
  elif age <= 25: cat = 'Student'
  elif age <= 35: cat = 'Young Adult'
  elif age <= 60: cat = 'Adult'
  else : cat = 'Elderly'

  return cat

# lambda 식에 위에서 생성한 get_category() 함수를 반환값으로 지정
# get_category(X)는 입력값으로 'Age' 칼럼 값을 받아서 해당하는 cat 반환
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : get_category(x))
titanic_df[['Age', 'Age_cat']].head()

Age      Age_cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult